src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+    return nullptr;
+  }
+  // addp: memory address for loading/storing an array element. It should be an
+  // AddP node operating on an array of specific type
+  Node* addp = mem->in(MemNode::Address);
+  if (!addp->is_AddP() || !operates_on_array_of_type(addp, mem_type)) {
+    return nullptr;
+  }
+  // Create a Node_Stack for SWPointer's initial stack
+  Node_Stack* nstack = new Node_Stack(_arena, 5);
+  nstack->push(addp, 0);
+  // addp2: another possible AddP node for array element addressing. It should
+  // operate on the same memory type and have the same base with previous AddP.
+  Node* addp2 = addp->in(AddPNode::Address);
+  if (addp2->is_AddP()) {
+    if (!operates_on_array_of_type(addp2, mem_type) ||
+        addp->in(AddPNode::Base) != addp2->in(AddPNode::Base)) {
+      return nullptr;
+    }
+    nstack->push(addp2, 1);
+  }
+
+  // Check supported memory access via SWPointer. It's not supported if
+  //  1) The constructed SWPointer is invalid
+  //  2) Address is growing down (index scale * loop stride < 0)
+  //  3) Memory access scale is different from data size
+  //  4) The loop increment node is on the SWPointer's node stack
+  SWPointer* ptr = new (_arena) SWPointer(mem, _phase, _lpt, nstack, true);
+  if (!ptr->valid()) {
+    return nullptr;
+  }
+  int scale_in_bytes = ptr->scale_in_bytes();
+  int element_size = type2aelembytes(mem_type);
+  if (scale_in_bytes * _cl->stride_con() < 0 ||
+      abs(scale_in_bytes) != element_size) {
+    return nullptr;
+  }
+  for (uint i = 0; i < nstack->size(); i++) {
+    if (nstack->node_at(i) == _cl->incr()) {
+      return nullptr;
+    }
+  }
+
+  return ptr;
+}
+
+// Check if node operates on an array of specific type
+bool VectorMaskedLoop::operates_on_array_of_type(Node* node, BasicType bt) {
+  const TypeAryPtr* aryptr = node->bottom_type()->isa_aryptr();
+  if (aryptr == nullptr) {
+    return false;
+  }
+  BasicType elem_bt = aryptr->elem()->array_element_basic_type();
+  return same_type_or_subword_size(elem_bt, bt);
+}
+
+// ------------------- Actual loop transformation functions -------------------
+// Create a tree of vector masks for use of vectorized operations in the loop
+Node_List* VectorMaskedLoop::create_vmask_tree(const TypeVectMask* t_vmask) {
+  // Create the root vector mask node from given vector type
+  int max_trip_cnt = _cl->trip_count();
+  Node* root_vmask = _cl->stride_con() > 0 ?
+      new LoopVectorMaskNode(_iv, _cl->limit(), t_vmask, max_trip_cnt) :
+      new LoopVectorMaskNode(_cl->limit(), _iv, t_vmask, max_trip_cnt);
+  _igvn->register_new_node_with_optimizer(root_vmask);
+
+  // Compute the depth of vector mask tree
+  uint small = _size_stats.smallest_size();
+  uint large = _size_stats.largest_size();
+  uint tree_depth = exact_log2(large) - exact_log2(small) + 1;
+  // All vector masks construct a perfect binary tree of "2 ^ depth - 1" nodes
+  // We create a list of "2 ^ depth" nodes for easier computation.
```

Comment:
Assume we have a small and a large type (byte and long). Size 1 and 8. `tree_depth = log2(8) - log2(1) + 1 = 3 - 0 + 1 = 4`. Then you generate a tree with `2^4-1 = 15` nodes. Did I calculate this right? That seems a bit excessive. Would be interesting to see benchmarks for mixed type cases.

Reply:
Can there be cases where creating the masks makes vectorization unprofitable?

Reply:
I have an example here:
```
public class Test {
    static int RANGE = 1024;

    public static void main(String[] strArr) {
        byte a[] = new byte[RANGE];
        long b[] = new long[RANGE];
        test0(a, b);
    }

    static void test0(byte[] a, long[] b) {
        for (int i = 0; i < RANGE; i++) {
            a[i]++;
            b[i]++;
        }
    }
}
```
`./java -Xcomp -XX:-TieredCompilation -XX:+TraceNewVectors -XX:+TraceLoopOpts -XX:+UnlockExperimentalVMOptions -XX:+UseMaskedLoop -XX:+TraceMaskedLoop -XX:CompileCommand=compileonly,Test::test0 Test.java`
This are the masks:
```
Generated vector masks in vmask tree
Lane_size = 1
 3710  LoopVectorMask  === _ 367 26  [[ 3711 3712 ]]  #vectormask[64]:{byte}
Lane_size = 2
 3711  ExtractLowMask  === _ 3710  [[ 3713 3714 ]]  #vectormask[32]:{short}
 3712  ExtractHighMask  === _ 3710  [[ 3715 3716 ]]  #vectormask[32]:{short}
Lane_size = 4
 3713  ExtractLowMask  === _ 3711  [[ 3717 3718 ]]  #vectormask[16]:{int}
 3714  ExtractHighMask  === _ 3711  [[ 3719 3720 ]]  #vectormask[16]:{int}
 3715  ExtractLowMask  === _ 3712  [[ 3721 3722 ]]  #vectormask[16]:{int}
 3716  ExtractHighMask  === _ 3712  [[ 3723 3724 ]]  #vectormask[16]:{int}
Lane_size = 8
 3717  ExtractLowMask  === _ 3713  [[ ]]  #vectormask[8]:{long}
 3718  ExtractHighMask  === _ 3713  [[ ]]  #vectormask[8]:{long}
 3719  ExtractLowMask  === _ 3714  [[ ]]  #vectormask[8]:{long}
 3720  ExtractHighMask  === _ 3714  [[ ]]  #vectormask[8]:{long}
 3721  ExtractLowMask  === _ 3715  [[ ]]  #vectormask[8]:{long}
 3722  ExtractHighMask  === _ 3715  [[ ]]  #vectormask[8]:{long}
 3723  ExtractLowMask  === _ 3716  [[ ]]  #vectormask[8]:{long}
 3724  ExtractHighMask  === _ 3716  [[ ]]  #vectormask[8]:{long}
```
That is indeed `15` masks. Hmm. Maybe that is the best one can do. And maybe it is not all that bad. But again, would be interesting to see the benchmarks for that case.

Reply:
We just benchmarked this "byte + long" case and saw some performance regressions after vectorization. Yes, too many mask operations are expensive. GCC does this in a better way: For adjacent data sizes (larger = 2 * smaller), it extracts two halves of the vector mask, but for non-adjacent data sizes, it re-generates vector masks without extraction.

Reply:
After some study, I know my current approach does create some unnecessary memory dependence. Will fix this after some refactoring work.